// durable-functions/lib/src/orchestrations/DurableOrchestrationStatus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DurableOrchestrationStatus = void 0;
class DurableOrchestrationStatus {
    constructor(init) {
        if (!this.isDurableOrchestrationStatusInit(init)) {
            throw new TypeError(`Failed to construct a DurableOrchestrationStatus object because the initializer had invalid types or missing fields. Initializer received: ${JSON.stringify(init)}`);
        }
        this.name = init.name;
        this.instanceId = init.instanceId;
        this.createdTime = new Date(init.createdTime);
        this.lastUpdatedTime = new Date(init.lastUpdatedTime);
        this.input = init.input;
        this.output = init.output;
        this.runtimeStatus = init.runtimeStatus;
        this.customStatus = init.customStatus;
        this.history = init.history;
    }
    isDurableOrchestrationStatusInit(obj) {
        const objAsInit = obj;
        return (objAsInit !== undefined &&
            typeof objAsInit.name === "string" &&
            typeof objAsInit.instanceId === "string" &&
            (typeof objAsInit.createdTime === "string" || objAsInit.createdTime instanceof Date) &&
            (typeof objAsInit.lastUpdatedTime === "string" ||
                objAsInit.lastUpdatedTime instanceof Date) &&
            objAsInit.input !== undefined &&
            objAsInit.output !== undefined &&
            typeof objAsInit.runtimeStatus === "string");
    }
}
exports.DurableOrchestrationStatus = DurableOrchestrationStatus;
